Source/WebCore/accessibility/AXTextMarker.cpp

```diff
@@ -902,28 +902,26 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
 
     // Functions to update resultMarker for word and sentence text units.
     auto updateWordResultMarker = [&] () {
-        int start, end;
-        findWordBoundary(flattenedRuns, offset, &start, &end);
-        if (direction == AXDirection::Previous) {
+        if (direction == AXDirection::Previous && boundary == AXTextUnitBoundary::Start) {
             int previousWordStart = findNextWordFromIndex(flattenedRuns, offset, false);
-            findEndWordBoundary(flattenedRuns, previousWordStart, &end);
-
-            // If start (from the forward-search) is the same as the offset, that means we are on a
-            // start word boundary and shouldn't update the text marker.
-            // When looking backward, the end of a word can be at the offset.
-            if (start != Checked<int>(offset) || end == Checked<int>(offset)) {
-                if (boundary == AXTextUnitBoundary::Start && previousWordStart < objectBorder && previousWordStart != -1)
-                    resultMarker = AXTextMarker(*currentObject, previousWordStart, origin);
-                else if (boundary == AXTextUnitBoundary::End && end <= objectBorder && end != -1)
-                    resultMarker = AXTextMarker(*currentObject, end, origin);
+            if (previousWordStart <= objectBorder)
+                resultMarker = AXTextMarker(*currentObject, previousWordStart, origin);
+        } else if (direction == AXDirection::Next && boundary == AXTextUnitBoundary::End) {
+            int nextWordEnd;
+            findEndWordBoundary(flattenedRuns, offset, &nextWordEnd);
+            if (nextWordEnd >= objectBorder) {
+                resultMarker = AXTextMarker(*currentObject, nextWordEnd - objectBorder, origin);
+                // Sometimes, the end word boundary will just return a whitespace word. For example: "Hello| world", with the text marker after hello, will return a text marker before world ("Hello |world"). Check for this case, and continue searching in that case.
```

Comment:
This is nice to have!

> Check for this case, and continue searching in that case.

Maybe could be written more simply, e.g. "If we detect this case, we want to continue searching for the next next-word-end" or something.